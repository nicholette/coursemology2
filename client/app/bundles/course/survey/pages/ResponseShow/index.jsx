import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { defineMessages, FormattedMessage } from 'react-intl';
import { browserHistory } from 'react-router';
import moment from 'moment';
import IconButton from 'material-ui/IconButton';
import { Card, CardText } from 'material-ui/Card';
import TitleBar from 'lib/components/TitleBar';
import Subheader from 'material-ui/Subheader';
import ArrowBack from 'material-ui/svg-icons/navigation/arrow-back';
import { Table, TableBody, TableRow, TableHeaderColumn, TableRowColumn } from 'material-ui/Table';
import surveyTranslations from 'course/survey/translations';
import { surveyShape, responseShape } from 'course/survey/propTypes';
import { fetchResponse } from 'course/survey/actions/responses';
import LoadingIndicator from 'course/survey/components/LoadingIndicator';
import ResponseForm from 'course/survey/containers/ResponseForm';
import RespondButton from 'course/survey/containers/RespondButton';
import UnsubmitButton from 'course/survey/containers/UnsubmitButton';

const translations = defineMessages({
  notSubmitted: {
    id: 'course.surveys.ResponseShow.notSubmitted',
    defaultMessage: 'Not submitted',
  },
});

const styles = {
  submissionInfoTable: {
    marginTop: 10,
    maxWidth: 600,
  },
};

class ResponseShow extends React.Component {
  static propTypes = {
    surveys: PropTypes.arrayOf(surveyShape),
    response: responseShape,
    flags: PropTypes.shape({
      isLoading: PropTypes.bool.isRequired,
      canModify: PropTypes.bool,
      canSubmit: PropTypes.bool,
    }),
    params: PropTypes.shape({
      courseId: PropTypes.string.isRequired,
      surveyId: PropTypes.string.isRequired,
      responseId: PropTypes.string.isRequired,
    }).isRequired,
    dispatch: PropTypes.func.isRequired,
  };

  componentDidMount() {
    const { dispatch, params: { responseId } } = this.props;
    dispatch(fetchResponse(responseId));
  }

  renderSubmissionInfo() {
    const { response } = this.props;
    return (
      <Table style={styles.submissionInfoTable}>
        <TableBody displayRowCheckbox={false}>
          <TableRow selectable={false}>
            <TableHeaderColumn>Student</TableHeaderColumn>
            <TableRowColumn>{response.creator_name}</TableRowColumn>
          </TableRow>
          <TableRow selectable={false}>
            <TableHeaderColumn>Submitted At</TableHeaderColumn>
            <TableRowColumn>
              {response.submitted_at ?
                moment(response.submitted_at).format('DD MMM YYYY, h:mma') :
                <FormattedMessage {...translations.notSubmitted} />
              }
            </TableRowColumn>
          </TableRow>
        </TableBody>
      </Table>
    );
  }

  renderBody() {
    const { response, flags } = this.props;
    if (flags.isLoading) { return <LoadingIndicator />; }

    return (
      <div>
        { this.renderSubmissionInfo() }
        <Subheader><FormattedMessage {...surveyTranslations.questions} /></Subheader>
        <ResponseForm
          readOnly
          initialValues={response}
          {...{ response, flags }}
        />
      </div>
    );
  }

  renderRespondButton(survey) {
    const { response, params: { courseId }, flags: { canModify, canSubmit, isLoading } } = this.props;
    if (!(canModify || canSubmit) || isLoading) { return null; }

    return (
      <RespondButton
        courseId={courseId}
        surveyId={survey.id}
        responseId={response.id}
        canModify={canModify}
        canSubmit={canSubmit}
        startAt={survey.start_at}
        endAt={survey.end_at}
        submittedAt={response.submitted_at}
      />
    );
  }

  renderUnsubmitButton() {
    const { response, flags: { canUnsubmit, isLoading } } = this.props;
    if (!canUnsubmit || isLoading) { return null; }
    return <UnsubmitButton responseId={response.id} />;
  }

  render() {
    const { surveys, params: { courseId, surveyId } } = this.props;
    const survey = surveys && surveys.length > 0 ?
                   surveys.find(s => String(s.id) === String(surveyId)) : {};

    return (
      <div>
        <TitleBar
          title={survey.title}
          iconElementLeft={<IconButton><ArrowBack /></IconButton>}
          onLeftIconButtonTouchTap={
            () => browserHistory.push(`/courses/${courseId}/surveys/${surveyId}`)
          }
        />
        { survey.description ? <Card><CardText>{survey.description}</CardText></Card> : null }
        { this.renderBody() }
        { this.renderRespondButton(survey) }
        { this.renderUnsubmitButton() }
      </div>
    );
  }
}

export const UnconnectedResponseShow = ResponseShow;

export default connect(
  state => ({
    ...state.responseForm,
    surveys: state.surveys,
  })
)(ResponseShow);
