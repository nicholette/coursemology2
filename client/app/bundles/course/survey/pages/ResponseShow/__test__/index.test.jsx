import React from 'react';
import { mount, shallow } from 'enzyme';
import { browserHistory } from 'react-router';
import CourseAPI from 'api/course';
import storeCreator from 'course/survey/store';
import ResponseShow, { UnconnectedResponseShow } from '../index';

const generateContextOptions = store => ({
  context: { intl, store, muiTheme },
  childContextTypes: { muiTheme: React.PropTypes.object, intl: intlShape },
});

describe('<ResponseShow />', () => {
  it('allows responses to be saved', async () => {
    const surveyId = '1';
    const responseId = '1';
    const responseUrl = `/courses/${courseId}/surveys/${surveyId}/responses/${responseId}`;
    const spyFetch = jest.spyOn(CourseAPI.survey.responses, 'fetch');

    // Mount response show page and wait for data to load
    Object.defineProperty(window.location, 'pathname', { value: responseUrl });
    mount(
      <ResponseShow params={{ courseId, surveyId, responseId }} />,
      generateContextOptions(storeCreator({}))
    );
    await sleep(1);
    expect(spyFetch).toHaveBeenCalled();
  });

  it('shows form and admin buttons if user has permissions and page is loaded', () => {
    const surveyId = 2;
    const responseId = 2;
    const surveys = [{
      id: surveyId,
      title: 'Survey',
      description: 'Description',
    }];
    const responseFormData = {
      response: {
        id: responseId,
        creator_name: 'Staff',
        submitted_at: '2099-12-31T16:00:00.000Z',
      },
      flags: {
        canModify: true,
        canSubmit: true,
        canUnsubmit: true,
        isResponseCreator: true,
        isLoading: false,
      },
    };
    const responseShow = shallow(
      <UnconnectedResponseShow
        params={{ courseId, surveyId: surveyId.toString(), responseId: responseId.toString() }}
        dispatch={() => {}}
        surveys={surveys}
        {...responseFormData}
      />, generateContextOptions(storeCreator({}))
    );
    expect(responseShow).toMatchSnapshot();
  });

  it('shows only title and loading indicator when loading', () => {
    const surveyId = 2;
    const responseId = 2;
    const responseFormData = {
      response: {
        id: responseId,
        creator_name: 'Student',
      },
      flags: {
        canModify: true,
        canSubmit: true,
        canUnsubmit: true,
        isResponseCreator: true,
        isLoading: true,
      },
    };
    const responseShow = shallow(
      <UnconnectedResponseShow
        params={{ courseId, surveyId: surveyId.toString(), responseId: responseId.toString() }}
        dispatch={() => {}}
        {...responseFormData}
      />, generateContextOptions(storeCreator({}))
    );
    expect(responseShow).toMatchSnapshot();
  });

  it('goes to SurveyShow page when title bar back button is triggered', () => {
    const surveyId = '3';
    const responseId = '3';

    const responseShow = shallow(
      <UnconnectedResponseShow
        params={{ courseId, surveyId, responseId }}
        dispatch={() => {}}
        flags={{ isLoading: true }}
      />, generateContextOptions(storeCreator({}))
    );

    browserHistory.push = jest.fn();
    responseShow.find('TitleBar').first().prop('onLeftIconButtonTouchTap')();
    expect(browserHistory.push).toHaveBeenCalledWith(`/courses/${courseId}/surveys/${surveyId}`);
  });
});
