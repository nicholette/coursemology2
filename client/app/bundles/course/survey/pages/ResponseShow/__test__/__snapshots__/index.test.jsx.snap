// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResponseShow /> shows form and admin buttons if user has permissions and page is loaded 1`] = `
<div>
  <TitleBar
    iconElementLeft={
      <IconButton
        disableTouchRipple={false}
        disabled={false}
        iconStyle={Object {}}
        tooltipPosition="bottom-center"
        touch={false}
      >
        <NavigationArrowBack />
      </IconButton>
    }
    onLeftIconButtonTouchTap={[Function]}
    title="Survey"
  />
  <Card
    expandable={false}
    expanded={null}
    initiallyExpanded={false}
  >
    <CardText>
      Description
    </CardText>
  </Card>
  <div>
    <Table
      allRowsSelected={false}
      fixedFooter={true}
      fixedHeader={true}
      height="inherit"
      multiSelectable={false}
      selectable={true}
      style={
        Object {
          "marginTop": 10,
          "maxWidth": 600,
        }
      }
    >
      <TableBody
        allRowsSelected={false}
        deselectOnClickaway={true}
        displayRowCheckbox={false}
        multiSelectable={false}
        preScanRows={true}
        selectable={true}
        style={Object {}}
      >
        <TableRow
          displayBorder={true}
          hoverable={false}
          hovered={false}
          selectable={false}
          selected={false}
          striped={false}
        >
          <TableHeaderColumn>
            Student
          </TableHeaderColumn>
          <TableRowColumn
            hoverable={false}
          >
            Staff
          </TableRowColumn>
        </TableRow>
        <TableRow
          displayBorder={true}
          hoverable={false}
          hovered={false}
          selectable={false}
          selected={false}
          striped={false}
        >
          <TableHeaderColumn>
            Submitted At
          </TableHeaderColumn>
          <TableRowColumn
            hoverable={false}
          >
            01 Jan 2100, 12:00am
          </TableRowColumn>
        </TableRow>
      </TableBody>
    </Table>
    <Subheader
      inset={false}
    >
      <FormattedMessage
        defaultMessage="Questions"
        id="course.surveys.questions"
        values={Object {}}
      />
    </Subheader>
    <InjectIntl(ReduxForm)
      flags={
        Object {
          "canModify": true,
          "canSubmit": true,
          "canUnsubmit": true,
          "isLoading": false,
          "isResponseCreator": true,
        }
      }
      initialValues={
        Object {
          "creator_name": "Staff",
          "id": 2,
          "submitted_at": "2099-12-31T16:00:00.000Z",
        }
      }
      readOnly={true}
      response={
        Object {
          "creator_name": "Staff",
          "id": 2,
          "submitted_at": "2099-12-31T16:00:00.000Z",
        }
      }
    />
  </div>
  <Connect(RespondButton)
    canModify={true}
    canSubmit={true}
    courseId="1"
    responseId={2}
    submittedAt="2099-12-31T16:00:00.000Z"
    surveyId={2}
  />
  <Connect(UnsubmitButton)
    responseId={2}
  />
</div>
`;

exports[`<ResponseShow /> shows only title and loading indicator when loading 1`] = `
<div>
  <TitleBar
    iconElementLeft={
      <IconButton
        disableTouchRipple={false}
        disabled={false}
        iconStyle={Object {}}
        tooltipPosition="bottom-center"
        touch={false}
      >
        <NavigationArrowBack />
      </IconButton>
    }
    onLeftIconButtonTouchTap={[Function]}
  />
  <LoadingIndicator />
</div>
`;
